### PR TITLE
InputGuard: Add guard class to get and verify input

### PR DESCRIPTION
**Pull request by pprkut**

### CODE REVIEW

File: src/Lunr/Flare/InputGuard.php [hunk 1]
```diff
@@ -0,0 +1,274 @@
...
+    /**
+     * Verify that a subject is set.
+     *
+     * @throws UninitializedException In case the InputGuard was initialized without an InputValue.
+     *
+     * @return void
+     */
+    protected function verifySubjectIsSet(): void
+    {
+        if (!isset($this->subject))
```
Comment (marcella-galambos-framna) [inline src/Lunr/Flare/InputGuard.php:150]: maybe we could add a test for not throwing an exception if the subject is null - since this is now different from the original stellr implementation

File: src/Lunr/Flare/InputGuard.php [hunk 2]
```diff
@@ -0,0 +1,274 @@
...
+            case RequestData::CliArgument:
+                // We JSON encode data here to make it show up like a list, even though the input isn't actually JSON
+                $rawValue = is_array($data) ? json_encode($data) : $data;
+                return $this->inspect($name, $data, $rawValue, $optional);
+            case RequestData::Upload:
+                // The original file name is probably the closest we get to the original input value,
+                // short of base64 encoding the actual file content
+                if (is_array($data) && isset($data['name']))
+                {
+                    $filename = is_array($data['name']) ? array_pop($data['name']) : $data['name'];
```
Comment (marcella-galambos-framna) [inline src/Lunr/Flare/InputGuard.php]: i'm missing a test case for the `array_pop` version

**Review by marcella-galambos-move — CHANGES_REQUESTED** [review]

File: src/Lunr/Flare/Tests/InputGuardBaseTest.php [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+    /**
+     * Test that the InputValue class was passed correctly.
+     */
+    public function testInputValuePassedCorrectly(): void
+    {
+        $this->assertPropertySame('subject', $this->subject);
+    }
+
+    /**
+     * Test that verifySubjectIsSet() does not throw an exception when subject is set.
```
Comment (marcella-galambos-framna) [inline src/Lunr/Flare/Tests/InputGuardBaseTest.php:71]:
```suggestion
     * Test that verifySubjectIsSet() does not throw an exception when subject is null.
```

File: src/Lunr/Flare/InputGuard.php [hunk 4]
```diff
@@ -0,0 +1,274 @@
...
+                    // The structure is determined by PHP internally. The filename can't be anything but a string.
+                    // The asserts helps phpstan take that into account.
+                    assert(is_string($filename));
+                }
+                else
+                {
+                    $filename = NULL;
+                }
+
+                return $this->inspect($name, $data, $filename, $optional);
```
Comment (marcella-galambos-framna) [inline src/Lunr/Flare/InputGuard.php:113]: i don't understand these changes now.

originally this is how it looks in Stellr right now:
`return $this->inspect($name, $data, $data['userfile']['name'], $optional);`

so now it can only be `$data['name']`, and it cannot be nested, like `$data['userfile']['name']`?

Comment (pprkut) [reply src/Lunr/Flare/InputGuard.php:113]: File upload is based on the `$_FILES` superglobal, which has a fixed structure:

```
Array
(
    [file2] => Array
    (
        [name] => MyFile.jpg
        [type] => image/jpeg
        [tmp_name] => /tmp/php/php6hst32
        [error] => UPLOAD_ERR_OK
        [size] => 98174
    )
)
```

Comments on https://www.php.net/manual/en/reserved.variables.files.php suggest that when you use filenames like `download[file1]` and `download[file2]` instead, you get a structure like this:

```

Array
(
    [download] => Array
        (
            [name] => Array
                (
                    [file1] => MyFile.txt
                    [file2] => MyFile.jpg
                )

            [type] => Array
                (
                    [file1] => text/plain
                    [file2] => image/jpeg
                )

            [tmp_name] => Array
                (
                    [file1] => /tmp/php/php1h4j1o
                    [file2] => /tmp/php/php6hst32
                )

            [error] => Array
                (
                    [file1] => UPLOAD_ERR_OK
                    [file2] => UPLOAD_ERR_OK
                )

            [size] => Array
                (
                    [file1] => 123
                    [file2] => 98174
                )
        )
)
```

`userfile` is the name of the uploaded filename. That was incorrectly hardcoded in the old, private code and now used as `$filename` instead

**Review by marcella-galambos-move — APPROVED** [review]
